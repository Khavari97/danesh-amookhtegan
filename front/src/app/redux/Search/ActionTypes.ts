export const SET_SEARCH_TEXT = "SET_SEARCH_TEXT"
export const SET_SEARCH_LOADING = "SET_SEARCH_LOADING"
export const SET_SEARCH_PROJECTS = "SET_SEARCH_PROJECTS"
export const SET_SEARCH_USERS = "SET_SEARCH_USERS"
export const SEARCH_PAGE = "SEARCH_PAGE"
export const SET_FILTER_CATEGORY_PROJECTS = "SET_FILTER_CATEGORY_PROJECTS"
export const SET_PROJECT_CATAGORY = "SET_PROJECT_CATAGORY"
export const SET_PROJECT_SKILL = "SET_PROJECT_SKILL"
export const SET_PROJECT_STATUS = "SET_PROJECT_STATUS"
export const SET_FILTER_USERS = "SET_FILTER__USERS"
export const SEARCH_USERS_PAGE = "SEARCH_USERS_PAGE"
export const SET_USER_UNIVERSITY = "SET_USER_UNIVERSITY"
export const SET_USER_PROVINCE = "SET_USER_PROVINCE"
export const SET_USER_CITY = "SET_USER_CITY"
export const SET_USER_SKILL = "SET_USER_SKILL"
export const GET_ALL_SKILLS = "GET_ALL_SKILLS"

export interface ISearchState {
    text: string,
    page: number,
    allSkills: [],
    countPages: number,
    loading: boolean,
    projects: [],
    users: [],
    project_category: string,
    project_status: string,
    project_skill: string,
    pageUsers: number,
    countUserPages: number,
    user_university: string,
    user_province: string,
    user_city: string,
    user_skill: string
}
export interface IGetAllSkills {
    type: typeof GET_ALL_SKILLS,
    allSkills: []
}

export interface ISetUserSkill {
    type: typeof SET_USER_SKILL,
    user_skill: string
}
export interface ISetUserCity {
    type: typeof SET_USER_CITY,
    user_city: string
}

export interface ISetUserProvince {
    type: typeof SET_USER_PROVINCE,
    user_province: string
}

export interface ISetUserUniversity {
    type: typeof SET_USER_UNIVERSITY,
    user_university: string
}

export interface ISetProjectCategory {
    type: typeof SET_PROJECT_CATAGORY,
    project_category: string
}
export interface ISetProjectSkill {
    type: typeof SET_PROJECT_SKILL,
    project_skill: string
}

export interface ISetProjectStatus {
    type: typeof SET_PROJECT_STATUS,
    project_status: string
}

export interface ISetSearchText {
    type: typeof SET_SEARCH_TEXT,
    text: string
}

export interface ISearchLoading {
    type: typeof SET_SEARCH_LOADING,
}

export interface ISearchProject {
    type: typeof SET_SEARCH_PROJECTS
    payload: {
        projects: []
    }
}

export interface ISearchUsers {
    type: typeof SET_SEARCH_USERS
    payload: {
        users: []
    }
}

export interface ISearchPage {
    type: typeof SEARCH_PAGE
    page: number
}

export interface ISetFilterCategoryProjects {
    type: typeof SET_FILTER_CATEGORY_PROJECTS
    projects: any,
    countPages: number
}

export interface ISetFilterUsers {
    type: typeof SET_FILTER_USERS
    users: any,
    countUserPages: number
}
export interface ISearchUsersPage {
    type: typeof SEARCH_USERS_PAGE
    pageUsers: number
}


export type ISearchActions = ISetSearchText
    | ISearchLoading | ISearchProject | ISearchUsers
    | ISearchPage | ISetFilterCategoryProjects
    | ISetProjectCategory | ISetProjectStatus | ISetFilterUsers
    | ISearchUsersPage | ISetUserUniversity
    | ISetUserProvince | ISetUserCity | IGetAllSkills
    | ISetUserSkill | ISetProjectSkill