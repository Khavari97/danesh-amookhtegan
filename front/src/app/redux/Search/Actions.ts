import { Dispatch } from "redux";
import { apiClient } from "../../api/_api";
import {
    SET_SEARCH_LOADING,
    SET_SEARCH_PROJECTS,
    SET_SEARCH_USERS,
    SET_FILTER_CATEGORY_PROJECTS,
    SEARCH_PAGE,
    SET_SEARCH_TEXT,
    SET_PROJECT_CATAGORY,
    SET_PROJECT_STATUS,
    SET_PROJECT_SKILL,
    SET_FILTER_USERS,
    SEARCH_USERS_PAGE,
    SET_USER_UNIVERSITY,
    SET_USER_PROVINCE,
    SET_USER_CITY,
    GET_ALL_SKILLS,
    SET_USER_SKILL
} from "./ActionTypes";


export const setProjectCategory = (category: string) => {
    return {
        type: SET_PROJECT_CATAGORY,
        project_category: category
    }
}

export const setProjectSkill = (skill: string) => {
    return {
        type: SET_PROJECT_SKILL,
        project_skill: skill
    }
}

export const setUserSkill = (skill: string) => {
    return {
        type: SET_USER_SKILL,
        user_skill: skill
    }
}
export const setUserUniversity = (university: string) => {
    return {
        type: SET_USER_UNIVERSITY,
        user_university: university
    }
}

export const setUserProvince = (province: string) => {
    return {
        type: SET_USER_PROVINCE,
        user_province: province
    }
}
export const setUserCity = (city: string) => {
    return {
        type: SET_USER_CITY,
        user_city: city
    }
}

export const setProjectStatus = (status: string) => {
    return {
        type: SET_PROJECT_STATUS,
        project_status: status
    }
}

export const setSearchText = (text: string) => {
    return {
        type: SET_SEARCH_TEXT,
        text: text
    }
}

export const setSearchLoading = () => {
    return {
        type: SET_SEARCH_LOADING,
    }
}


export const setSearchProjects = (projects: []) => {
    return {
        type: SET_SEARCH_PROJECTS,
        payload: {
            projects: projects
        }
    }
}

export const setSearchUsers = (users: []) => {
    return {
        type: SET_SEARCH_USERS,
        payload: {
            users: users
        }
    }
}


export const searchPage = (page: number) => {
    return {
        type: SEARCH_PAGE,
        page: page
    }
}


export const searchUsersPage = (page: number) => {
    return {
        type: SEARCH_USERS_PAGE,
        pageUsers: page
    }
}

export const Filter = (status: string, category: string, skill: string, text: string, page: number) => (
    dispatch: Dispatch
) => {
    try {
        apiClient
            .get(`dashboard/projects/?search=${text}&category=${category}&skill=${skill}&status=${status}&page=${page}`, {
                consoleMessage: true,
            })
            .then((response) => {
                dispatch({
                    type: SET_FILTER_CATEGORY_PROJECTS,
                    projects: response.results,
                    countPages: Math.ceil(response.count / 6)
                });
            })
            .catch((error) => {
            });
    } catch {
    }
}

export const FilterUsers = (university: string, skill: string, city: string, province: string, searchText: string, page: number) => (
    dispatch: Dispatch
) => {
    try {
        apiClient
            .get(`dashboard/users/?search=${searchText}&university=${university}&skill=${skill}&city=${city}&province=${province}&page=${page}`, {
                consoleMessage: true,
            })
            .then((response) => {
                console.log(Math.ceil(response.count / 6))
                dispatch({
                    type: SET_FILTER_USERS,
                    users: response.results,
                    countUserPages: Math.ceil(response.count / 6)
                });
            })
            .catch((error) => {
            });
    } catch {
    }
}

export const getAllSkills = () => (
    dispatch: Dispatch
) => {
    try {
        apiClient
            .get(`list/skills/?all=true`, {
                consoleMessage: true,
            })
            .then((response) => {
                dispatch({
                    type: GET_ALL_SKILLS,
                    allSkills: response,
                });
            })
            .catch((error) => {
            });
    } catch {
    }
}

