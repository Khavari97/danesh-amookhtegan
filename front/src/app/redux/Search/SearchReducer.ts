import {
    ISearchActions,
    ISearchState,
    SEARCH_PAGE,
    SET_SEARCH_LOADING,
    SET_SEARCH_PROJECTS,
    SET_SEARCH_TEXT,
    SET_SEARCH_USERS,
    SET_FILTER_CATEGORY_PROJECTS,
    SET_PROJECT_CATAGORY,
    SET_PROJECT_STATUS,
    SET_FILTER_USERS,
    SEARCH_USERS_PAGE,
    SET_USER_UNIVERSITY,
    SET_USER_PROVINCE,
    SET_USER_CITY,
    GET_ALL_SKILLS,
    SET_USER_SKILL,
    SET_PROJECT_SKILL
} from "./ActionTypes";

const initialState: ISearchState = {
    text: "",
    page: 1,
    allSkills: [],
    countPages: 0,
    loading: true,
    projects: [],
    users: [],
    project_category: '',
    project_status: '',
    project_skill: '',
    pageUsers: 1,
    countUserPages: 0,
    user_university: '',
    user_province: '',
    user_city: '',
    user_skill: ''
}

export const SearchReducer = (state = initialState, action: ISearchActions): ISearchState => {
    switch (action.type) {
        case SET_SEARCH_TEXT:
            return {
                ...state,
                text: action.text,
                page:1,
                pageUsers:1
            }
        case SET_SEARCH_PROJECTS:
            return {
                ...state,
                loading: false,
                projects: action.payload.projects
            }
        case SET_SEARCH_USERS:
            return {
                ...state,
                loading: false,
                users: action.payload.users
            }
        case SET_SEARCH_LOADING:
            return {
                ...state,
                loading: true
            }
        case SEARCH_PAGE:
            return {
                ...state,
                loading: false,
                page: action.page
            }
        case GET_ALL_SKILLS:
            return {
                ...state,
                allSkills: action.allSkills,
            }
        case SET_FILTER_CATEGORY_PROJECTS:
            return {
                ...state,
                loading: false,
                projects: action.projects,
                countPages: action.countPages
            }
        case SET_PROJECT_CATAGORY:
            return {
                ...state,
                loading: false,
                project_category: action.project_category,
                page: 1
            }
        case SET_PROJECT_STATUS:
            return {
                ...state,
                loading: false,
                project_status: action.project_status,
                page: 1
            }
        case SET_FILTER_USERS:
            return {
                ...state,
                loading: false,
                users: action.users,
                countUserPages: action.countUserPages
            }
        case SEARCH_USERS_PAGE:
            return {
                ...state,
                loading: false,
                pageUsers: action.pageUsers
            }
        case SET_USER_UNIVERSITY:
            return {
                ...state,
                loading: false,
                user_university: action.user_university,
                pageUsers: 1
            }
        case SET_USER_PROVINCE:
            return {
                ...state,
                loading: false,
                user_province: action.user_province,
                pageUsers: 1
            }
        case SET_USER_CITY:
            return {
                ...state,
                loading: false,
                user_city: action.user_city,
                pageUsers: 1
            }
        case SET_USER_SKILL:
            return {
                ...state,
                loading: false,
                user_skill: action.user_skill,
                pageUsers: 1
            }
        case SET_PROJECT_SKILL:
            return {
                ...state,
                loading: false,
                project_skill: action.project_skill,
                page: 1
            }
        default:
            return {
                ...state
            }
    }
}