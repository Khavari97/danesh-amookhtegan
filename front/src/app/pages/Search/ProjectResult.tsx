import React, { Component } from 'react';
import Loading from "../../components/customs/Loading/Loading";
import { Grid } from "../../components/material-ui";
import ProjectNotFound from "../../components/customs/ProjectNotFound/ProjectNotFound";
import { apiClient } from "../../api/_api";
import ProjectCard from "../../components/customs/ProjectCard/ProjectCard";
import { IRootState } from "../../redux/RootReducer";
import { connect } from "react-redux";
import { Filter, searchPage, setSearchText } from "../../redux/Search/Actions";
import { ISearchState } from "../../redux/Search/ActionTypes";
import { withRouter } from "react-router-dom";
import SearchProjectFilter from "./SearchProjectFilter";
import BasicPagination from "../../components/customs/Pagination/BasicPagination";
import SearchInput from "../../components/customs/SearchInput/SearchInput";


const mapStateToProps = (store: IRootState) => {
    return store.search
}

const mapDispatchToProps = (dispatch: any) => {
    return {
        FilterProjects: (status: string, category: string, skill: string, text: string, countPages: number) => dispatch(Filter(status, category, skill, text, countPages)),
        searchPage: (page: number) => dispatch(searchPage(page)),
        setSearchText: (text: string) => dispatch(setSearchText(text)),
    }
}
interface IState {
    data: any;
    page: any;
    number: any;
    text: any;
}

interface IProps extends ISearchState {
    location: any,
    FilterProjects: any;
    searchPage: any;
    setSearchText: any
}

class ProjectResult extends Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);
        this.state = {
            data: [],
            page: 1,
            number: 1,
            text: ""

        }
    }

    handleChange = (event: React.ChangeEvent<unknown>, value: number) => {
        this.props.searchPage(value)
    };


    render() {

        return (
            <>
                <SearchInput data_testid="search-input" text={(event: any) => {
                    if (event.target.value.length >= 3 || event.target.value.length === 0) {
                        this.props.setSearchText(event.target.value)
                    }
                }}
                    default={this.props.text}
                />

                <Grid>
                    <SearchProjectFilter />
                </Grid>

                {
                    this.props.loading ? <Loading /> :
                        <Grid className='animate__animated animate__fadeIn' container spacing={2}>
                            {this.props.projects?.length !== 0 ?
                                this.props.projects?.map((project: any, index: number) => {
                                    return <Grid data-testid="project-card" item xs={12} sm={6} md={4}
                                        key={index}><ProjectCard {...project} /></Grid>
                                }) : <ProjectNotFound />}
                        </Grid>
                }
                <BasicPagination siblingCount={0} boundaryCount={2} page={this.props.page} count={this.props.countPages} onChange={this.handleChange} />
            </>
        );
    }

    componentDidMount() {
        this.props.searchPage(1)
    }

    componentDidUpdate(prevProps: IProps) {
        if (this.props.text !== prevProps.text) {
            this.props.FilterProjects(this.props.project_status, this.props.project_category, this.props.project_skill, this.props.text, this.props.page)
        }
        if (this.props.page !== prevProps.page) {
            this.props.FilterProjects(this.props.project_status, this.props.project_category, this.props.project_skill, this.props.text, this.props.page)
        }
    }


}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ProjectResult));
