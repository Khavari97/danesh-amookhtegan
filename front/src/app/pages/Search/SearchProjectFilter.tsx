import React, { Component } from 'react';
import { rootState } from "../../redux/RootReducer";
import { GetParentSkills, GetSkills, } from "../../redux/User/Profile/Actions";
import { connect } from "react-redux";
import { Grid } from "@material-ui/core";
import Autocomplete from "@material-ui/lab/Autocomplete";
import TextField from "@material-ui/core/TextField";
import { Filter, getAllSkills, setProjectCategory, setProjectSkill, setProjectStatus } from '../../redux/Search/Actions';

interface IState {
    fields: any,
    filter: { category: string, status: string, skill: string }
}

interface IProps {
    status: any;
    parentSkill: any;
    childSkill: any;
    GetSkills: any;
    GetParentSkills: any;
    FilterProjects: any;
    setProjectCategory: any;
    page: number;
    project_category: string;
    project_skill: string;
    project_status: string;
    setProjectStatus: any;
    text: string;
    getAllSkills: any;
    allSkills: [];
    setProjectSkill: any

}

const mapStateToProps = (state: rootState) => {
    return {
        status: state.project?.status,
        parentSkill: state.userProfile?.parentSkills,
        childSkill: state.userProfile?.skills,
        countPages: state.search?.countPages,
        page: state.search?.page,
        project_category: state.search?.project_category,
        project_skill: state.search?.project_skill,
        project_status: state.search?.project_status,
        text: state.search?.text,
        allSkills: state.search?.allSkills
    }
}

const mapDispatchToProps = (dispatch: any) => {
    return {
        GetSkills: () => dispatch(GetSkills()),
        GetParentSkills: () => dispatch(GetParentSkills()),
        FilterProjects: (status: string, category: string, skill: string, text: string, countPages: number) => dispatch(Filter(status, category, skill, text, countPages)),
        setProjectCategory: (category: string) => dispatch(setProjectCategory(category)),
        setProjectStatus: (status: string) => dispatch(setProjectStatus(status)),
        setProjectSkill: (skill: string) => dispatch(setProjectSkill(skill)),
        getAllSkills: () => dispatch(getAllSkills())
    }
}

class SearchProjectFilter extends Component<IProps, IState>{
    constructor(props: IProps) {
        super(props);
        this.state = {
            filter: { status: '', category: '', skill: '' },
            fields: [
                {
                    id: 'selectInput',
                    type: 'text',
                    name: ' status ',
                    label: ' وضعیت ',
                    value: ' ',
                    items: []
                },
                {
                    id: 'selectInput',
                    type: 'text',
                    name: ' parentSkill ',
                    label: ' دسته بندی ',
                    value: ' ',
                    items: props.parentSkill
                },
                {
                    id: 'selectInput',
                    type: 'text',
                    name: ' childSkill ',
                    label: ' مهارت ',
                    value: ' ',
                    items: props.childSkill
                }
            ]

        }
    }

    onChangeHandler = (event: any, values: any) => {
        console.log(values)
        this.props.setProjectCategory((values?.name) ? values.name : '')
    }

    render() {
        return (
            <Grid container>
                <Grid item data-test="sorting" style={{ padding: "0 5px 15px 0" }} sm={12}> دسته بندی بر اساس : </Grid>
                {
                    this.state.fields.map((field: any,index:number) => {
                        return (
                            <Grid key={index} item sm={4} xs={12}>
                                <Autocomplete
                                    id="combo-box-demo"
                                    data-test="Autocomplete"
                                    options={field.items}
                                    getOptionLabel={(option: any) => option.name}
                                    onChange={field.handleChange}
                                    size="small"
                                    style={{
                                        padding: "5px 5px",
                                        marginBottom: "20px"
                                    }}
                                    data-testid={field.name}
                                    renderInput={(params) =>
                                        <TextField {...params} label={field.label} variant="outlined" data-test="field" />}
                                />

                            </Grid>
                        )
                    })
                }

            </Grid>

        );
    }

    componentDidMount() {
        this.props.GetSkills();
        this.props.GetParentSkills();
        this.props.getAllSkills();
        this.props.FilterProjects('', '', '', this.props.text, this.props.page)
    }

    componentDidUpdate(prevProps: IProps) {
        if (this.props.project_category !== prevProps.project_category) {
            this.props.FilterProjects(this.props.project_status, this.props.project_category, this.props.project_skill, this.props.text, this.props.page)
        }
        if (this.props.project_status !== prevProps.project_status) {
            this.props.FilterProjects(this.props.project_status, this.props.project_category, this.props.project_skill, this.props.text, this.props.page)
        }
        if (this.props.project_skill !== prevProps.project_skill) {
            this.props.FilterProjects(this.props.project_status, this.props.project_category, this.props.project_skill, this.props.text, this.props.page)
        }
    }

    static getDerivedStateFromProps(nextProps: IProps, nextState: IState) {
        return {
            fields: [

                {
                    id: 'selectInput',
                    type: 'text',
                    name: ' status ',
                    label: ' وضعیت ',
                    value: ' ',
                    handleChange: (event: any, values: any) => {
                        nextProps.setProjectStatus((values?.code) ? values.code : '')
                    },
                    items: [
                        {
                            code: "STARTED",
                            name: "شروع شده"
                        },
                        {
                            code: "ENDED",
                            name: "پایان یافته"
                        },
                        {
                            code: "WAITING",
                            name: "در حال انتظار"
                        },
                        {
                            code: "DELETED",
                            name: "حذف شده"
                        }
                    ]
                },

                {
                    id: 'selectInput',
                    type: 'text',
                    name: ' parentSkill ',
                    label: ' دسته بندی ',
                    value: ' ',
                    handleChange: (event: any, values: any) => {
                        nextProps.setProjectCategory((values?.name) ? values.name : '')
                    },
                    items: nextProps.parentSkill
                },
                {
                    id: 'selectInput',
                    type: 'text',
                    name: ' childSkill ',
                    label: ' مهارت ',
                    value: ' ',
                    items: nextProps.allSkills,
                    handleChange: (event: any, values: any) => {
                        nextProps.setProjectSkill((values?.name) ? values.name : '')
                    },
                }
            ]

        }
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(SearchProjectFilter)