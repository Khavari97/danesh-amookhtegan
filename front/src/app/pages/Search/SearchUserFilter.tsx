import React, { Component } from "react";
import { rootState } from "../../redux/RootReducer";
import { GetSkills, ProvinceApi } from "../../redux/User/Profile/Actions";
import { connect } from "react-redux";
import { Grid } from '@material-ui/core'
import TextField from '@material-ui/core/TextField';
import Autocomplete from '@material-ui/lab/Autocomplete';
import { FilterUsers, getAllSkills, setUserCity, setUserProvince, setUserSkill, setUserUniversity } from "../../redux/Search/Actions";

interface IProps {
    skills: any;
    university: any;
    province: any;
    city: any;
    GetSkills: any
    ProvinceApi: any
    countUserPages: any
    FilterUsers: any
    pageUsers: number
    setUserUniversity: any
    user_university: string
    setUserProvince: any
    setUserSkill: any
    user_province: string
    user_city: string
    user_skill: string
    setUserCity: any
    text: string
    getAllSkills: any
    allSkills: []
    // FilterCity:any
}

interface IState {
    fields: any
}

const mapStateToProps = (state: rootState) => {
    return {
        skills: state.userProfile.skills,
        university: state.userProfile.universities,
        province: state.userProfile.provinces,
        city: state.userProfile.cities,
        countUserPages: state.search.countUserPages,
        pageUsers: state.search.pageUsers,
        user_university: state.search.user_university,
        user_province: state.search.user_province,
        user_city: state.search.user_city,
        text: state.search.text,
        allSkills: state.search.allSkills,
        user_skill: state.search.user_skill,
    }
}

const mapDispatchToProps = (dispatch: any) => {
    return {
        ProvinceApi: () => dispatch(ProvinceApi()),
        GetSkills: () => dispatch(GetSkills()),
        FilterUsers: (university: string, skill: string, city: string, province: string, text: string, countPages: number) => dispatch(FilterUsers(university, skill, city, province, text, countPages)),
        setUserUniversity: (university: string) => dispatch(setUserUniversity(university)),
        setUserProvince: (province: string) => dispatch(setUserProvince(province)),
        setUserCity: (city: string) => dispatch(setUserCity(city)),
        getAllSkills: () => dispatch(getAllSkills()),
        setUserSkill: (skill: string) => dispatch(setUserSkill(skill)),
    }
}


class SearchUserFilter extends Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);
        this.state = {
            fields: [

                {
                    id: 'selectInput',
                    name: 'skills',
                    label: ' مهارت ',
                    value: '',
                    placeholder: '',
                    items: props.skills

                }
                ,

                {
                    id: 'selectInput',
                    name: 'university',
                    label: ' دانشگاه ',
                    value: '',
                    placeholder: '',
                    items: props.university

                }
                ,

                {
                    id: 'selectInput',
                    name: 'province',
                    label: ' استان ',
                    value: '',
                    code: '',
                    placeholder: '',
                    items: props.province
                },
                {
                    id: 'selectInput',
                    name: 'city',
                    label: ' شهر',
                    value: '',
                    placeholder: '',
                    items: props.city
                }

            ]

        }
    }


    render() {

        return (
            <Grid container>
                <Grid item data-test="sorting" style={{ padding: "0 5px 15px 0", width: "100%" }} sm={12}> دسته بندی بر اساس : </Grid>
                {
                    this.state.fields.map((field: any,index:number) => {
                        return (
                            <Grid key={index} item sm={3} xs={12} >
                                <Autocomplete
                                   data-testid={field.name}
                                    id="combo-box-demo"
                                    options={field.items}
                                    getOptionLabel={(option: any) => option.name}
                                    size="small"
                                    style={{
                                        padding: "5px 3px",
                                        marginBottom: "20px"
                                    }}
                                    onChange={field.handleChange}
                                    renderInput={(params) =>
                                        <TextField {...params} label={field.label} variant="outlined" />}
                                />
                            </Grid>
                        )
                    })
                }

            </Grid>
        );
    }

    componentDidMount() {
        this.props.ProvinceApi();
        this.props.getAllSkills();
        this.props.FilterUsers(this.props.user_university, this.props.user_skill, this.props.user_city, this.props.user_province, this.props.text, this.props.pageUsers)
    }

    componentDidUpdate(prevProps: IProps) {
        if (this.props.user_university !== prevProps.user_university) {
            this.props.FilterUsers(this.props.user_university, this.props.user_skill, this.props.user_city, this.props.user_province, this.props.text, this.props.pageUsers)
        }
        if (this.props.user_province !== prevProps.user_province) {
            this.props.FilterUsers(this.props.user_university, this.props.user_skill, this.props.user_city, this.props.user_province, this.props.text, this.props.pageUsers)
        }
        if (this.props.user_city !== prevProps.user_city) {
            this.props.FilterUsers(this.props.user_university, this.props.user_skill, this.props.user_city, this.props.user_province, this.props.text, this.props.pageUsers)
        }

        if (this.props.text !== prevProps.text) {
            this.props.FilterUsers(this.props.user_university, this.props.user_skill, this.props.user_city, this.props.user_province, this.props.text, this.props.pageUsers)
        }
        if (this.props.user_skill !== prevProps.user_skill) {
            this.props.FilterUsers(this.props.user_university, this.props.user_skill, this.props.user_city, this.props.user_province, this.props.text, this.props.pageUsers)
        }
    }

    static getDerivedStateFromProps(nextProps: IProps, nextState: IState) {
        return {
            fields: [
                {
                    id: 'selectInput',
                    name: 'skills',
                    label: ' مهارت ',
                    value: '',
                    placeholder: '',
                    items: nextProps.allSkills,
                    handleChange: (event: any, values: any) => {
                        nextProps.setUserSkill((values?.id) ? values.id : '')
                    },
                }
                ,
                {
                    id: 'selectInput',
                    name: 'university',
                    label: ' دانشگاه ',
                    value: '',
                    placeholder: '',
                    items: nextProps.university,
                    handleChange: (event: any, values: any) => {
                        nextProps.setUserUniversity((values?.code) ? values.code : '')
                    },
                }
                ,
                {
                    id: 'selectInput',
                    name: 'province',
                    label: ' استان ',
                    value: '',
                    placeholder: '',
                    items: nextProps.province,
                    handleChange: (event: any, values: any) => {
                        nextProps.setUserProvince((values?.code) ? values.code : '')
                    },
                },
                {
                    id: 'selectInput',
                    name: 'city',
                    label: ' شهر',
                    value: '',
                    placeholder: '',
                    items: nextProps.city,
                    handleChange: (event: any, values: any) => {
                        nextProps.setUserCity((values?.code) ? values.code : '')
                    },
                }
            ]
        }
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(SearchUserFilter)
