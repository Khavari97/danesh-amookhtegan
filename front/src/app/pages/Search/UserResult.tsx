import React, { Component } from 'react';
import Loading from "../../components/customs/Loading/Loading";
import { Grid } from "../../components/material-ui";
import MemberCard from "../../components/customs/MemberCard/MemberCard";
import UserNotFound from "../../components/customs/UserNotFound/UserNotFound";
import { IRootState } from "../../redux/RootReducer";
import { FilterUsers, searchUsersPage, setSearchText } from "../../redux/Search/Actions";
import { connect } from "react-redux";
import { ISearchState } from "../../redux/Search/ActionTypes";
import { withRouter } from "react-router-dom";
import SearchUserFilter from "./SearchUserFilter";
import BasicPagination from "../../components/customs/Pagination/BasicPagination";
import SearchInput from "../../components/customs/SearchInput/SearchInput";


const mapStateToProps = (store: IRootState) => {
    return store.search
}

const mapDispatchToProps = (dispatch: any) => {
    return {
        searchUsersPage: (page: number) => dispatch(searchUsersPage(page)),
        setSearchText: (text: string) => dispatch(setSearchText(text)),
        FilterUsers: (university: string, skill: string, city: string, province: string, text: string, countPages: number) => dispatch(FilterUsers(university, skill, city, province, text, countPages)),
    }
}

interface IState {
    data: any;
    page: any;
    number: any;
    text: any;
    search: any
}

interface IProps extends ISearchState {
    location: any
    searchUsersPage: any
    FilterUsers: any
    setSearchText: any
}


class UserResult extends Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);
        this.state = {
            data: [],
            page: 1,
            number: 1,
            text: "",
            search: []
        }
    }

    handleChange = (event: React.ChangeEvent<unknown>, value: number) => {
        this.props.searchUsersPage(value)
    };

    render() {
        return (
            <>
                <SearchInput data_testid='search-input' text={(event: any) => {
                    if (event.target.value.length >= 3 || event.target.value.length === 0) {
                        this.props.setSearchText(event.target.value)
                    }
                }}
                    default={this.props.text}
                />

                <Grid>
                    <SearchUserFilter />
                </Grid>
                {
                    this.props.loading ? <Loading /> :
                        <Grid className='animate__animated animate__fadeIn' container spacing={2}>
                            {this.props.users?.length !== 0 ?
                                this.props.users.map((user: any, index: number) => {
                                    return <Grid data-testid="card_user" item xs={12} sm={6} md={4} key={index}>
                                        <MemberCard {...user} /></Grid>
                                }) : <UserNotFound />}

                        </Grid>
                }
                <BasicPagination siblingCount={0} boundaryCount={2} page={this.props.pageUsers} count={this.props.countUserPages} onChange={this.handleChange} />
            </>
        );
    }
    componentDidMount() {
        this.props.searchUsersPage(1)
    }
    componentDidUpdate(prevProps: IProps) {
        if (this.props.pageUsers !== prevProps.pageUsers) {
            this.props.FilterUsers(this.props.user_university, this.props.user_skill, this.props.user_city, this.props.user_province, this.props.text, this.props.pageUsers)
        }
    }
}


export default withRouter(connect(mapStateToProps, mapDispatchToProps)(UserResult));