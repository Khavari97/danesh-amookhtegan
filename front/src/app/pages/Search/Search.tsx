import React, { Component } from 'react';
import { Container, Typography } from "@material-ui/core";
import CustomTabs, { TabBarProps } from "../../components/customs/CustomTabs/CustomTabs";
import ProjectResult from "./ProjectResult";
import UserResult from "./UserResult";
import SearchInput from "../../components/customs/SearchInput/SearchInput";


class Search extends Component<any> {

    SearchTabs: TabBarProps[] = [
        { label: "پروژه", component: <ProjectResult /> },
        { label: "افراد", component: <UserResult /> },
    ]

    render() {
        return (
            <Container>
                {this.props.homeVersion ? "" : <Typography data-testid="h1" variant={"h1"} component={"h1"}>جستجو</Typography>}
                <CustomTabs tabs={this.SearchTabs} />
            </Container>
        )
    }
}


export default Search;