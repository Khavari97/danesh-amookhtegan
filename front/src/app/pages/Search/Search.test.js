import React from 'react';
import Enzyme, { shallow } from "enzyme";
import EnzymeAdapter from "enzyme-adapter-react-16";
import { Provider } from "react-redux";
import store from '../../redux/store';
import Search from './Search';
import renderer from "react-test-renderer";
import { render, screen, prettyDOM, act, waitFor, wait, fireEvent, queryByAttribute } from '@testing-library/react'
import ProjectResult from "./ProjectResult";
import UserResult from "./UserResult";
import { BrowserRouter } from 'react-router-dom'
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'
const middlewares = [thunk]
const mockStore = configureStore(middlewares);

Enzyme.configure({ adapter: new EnzymeAdapter() });
const setup = () => {
  return shallow(
    <Provider store={store}>
      <Search />
    </Provider>

  );
};

describe('render search component without error', () => {

  test("render test", () => {
    let wrapper = setup();
    expect(wrapper).toBeTruthy();
    expect(wrapper.length).toEqual(1);
  })

  test('snapshot', () => {
    const snapshot = renderer.create(setup).toJSON();
    expect(snapshot).toMatchSnapshot();
  })
})




describe('integration component', () => {
  const getById = queryByAttribute.bind(null, 'id');

  let mock_store

  beforeEach(() => {
    mock_store = mockStore({
    search: {
      text:""
      },
    });
  });

  test("render test", () => {

    const SearchTabs = [
      { label: "پروژه", component: <ProjectResult />, testid: "proje" },
      { label: "افراد", component: <UserResult />, testid: "users" },
    ]


    const dom = render(
      <BrowserRouter>
        <Provider store={store}>
          <Search SearchTabs={SearchTabs} />
        </Provider>
      </BrowserRouter>
    );

    fireEvent.click(screen.getAllByRole('tab')[1])
    expect(getById(dom.container, 'simple-tab-1')).toBeTruthy()


  })

  test("render test", () => {

    const SearchTabs = [
      { label: "پروژه", component: <ProjectResult />, testid: "proje" },
      { label: "افراد", component: <UserResult />, testid: "users" },
    ]


    const dom = render(
      <BrowserRouter>
        <Provider store={store}>
          <Search SearchTabs={SearchTabs} />
        </Provider>
      </BrowserRouter>
    );

    fireEvent.click(screen.getAllByRole('tab')[0])
    expect(getById(dom.container, 'simple-tab-0')).toBeTruthy()


  })

})


