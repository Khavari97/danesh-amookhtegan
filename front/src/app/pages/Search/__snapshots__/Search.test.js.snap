// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render search component without error snapshot 1`] = `null`;
